### ISSUE TITLE
Moving window computations

### ISSUE THREAD

**Issue by cscherrer** [tkf/Transducers.jl#39]
Hi,

I'm getting some strange behavior trying to do computations that work with a moving window. I'm guessing there's a stateful computation in here somewhere messing things up, but I'm not sure where. What's the right way to do this?

```julia
julia> xf = Zip(Map(identity), Drop(1)) |> MapSplat(-)
Map(_zip_init) |>
    TeeZip(
        Map(first) |>
            Map(identity)
    ) |>
    Map(_zip_between) |>
    TeeZip(
        Map(first) |>
            Drop(1)
    ) |>
    Map(_zip_between) |>
    Map(last) |>
    MapSplat(-)

julia> collect(xf, [2,3,5,7,11])
4-element Array{Int64,1}:
 0
 0
 0
 0
```

**Comment by tkf** [comment]
If you just want a moving window, you can use `Partition`:
https://tkf.github.io/Transducers.jl/dev/manual/#Transducers.Partition

However, `Partition` is too general and not super efficient.  In particular, it's written in a way that the input can be, e.g., filtered by other transducers.  But I think moving window is almost always the first stage of the processing pipeline, right?  In that case, operating on a vector directly is almost always the right thing to do because you can just take a view of it.

I think this kind of computation is better to do at the `foldl` side.  If you are interested, have a look at this `circularwindows` example in <https://tkf.github.io/Transducers.jl/dev/manual/#Transducers.AdHocFoldable>. I've been thinking how to build some consistent way to do this #7 but I still don't know the answer.  Maybe this kind of interface should be built on top of `foldl`/`reduce` and transducers (maybe something like Query.jl), not building in Transducers.jl.

FYI, the behavior of `Zip(Map(identity), Drop(1))` is actually expected and briefly "mentioned" in the documentation

> Head transducers drive tail transducers. Be careful when using it with transducers other than Map, especially the contractive ones like PartitionBy and the expansive ones like MapCat.
>
> --- https://tkf.github.io/Transducers.jl/dev/manual/#Transducers.Zip-Tuple

Yes, I know it's very cryptic and I have to find a better way to explain this other than saying read the code.  In `Zip(Map(identity), Drop(1))`, transducer `Drop(1)` "gates" the output from `Map(identity)`.  You can probably see it from the ugly tree printed in the REPL you quoted; `Map(identity)` comes before `Drop(1)` in the sequence of `|>`s.  This is actually useful with `Filter`:

```julia
julia> collect(Zip(Filter(isodd), Map(string)) |> MapSplat(=>), 1:5)
3-element Array{Pair{Int64,String},1}:
 1 => "1"
 3 => "3"
 5 => "5"
```

`Zip` in Transducers.jl is more about fan-out than zipping so maybe it's just that I chose really bad name.  Maybe I'll rename it to `Tee` or something.

---

tl;dr: At the moment, use:

* `eachcol` + `reshape` if you need performance.
* `Iterators.partition` if you need to handle edge cases and window size is large (It yields vectors).
* `IterTools.partition` if you only need short windows (it yields tuples).
* `Transducers.Partition` if you need some pre-processing (it's probably not so bad in this case)

**Comment by cscherrer** [comment]
>  But I think moving window is almost always the first stage of the processing pipeline, right?

Not in this case. As a starting point, I was doing this:
```julia
function Elbo(p,q)
    sample = Map(a -> rand(q))
    logp = Map(x -> logpdf(p,x))
    logq = Map(x -> logpdf(q,x))
    sample |> Zip(logp,logq) |> MapSplat(-) |> Transducer(Mean())
end

function elbo(p,q)
    xf = Elbo(p,q) # |> Converged()
    mapfoldl(xf, right, 1:1000)
end

f(σ) = elbo(Cauchy(),Normal(0,σ))
```

Here `p` and `q` are `Distribution`s (or `Model`s from Soss, but that's not important). 

Rather than fix a number of iterations, I was thinking it could be nice to have some general-purpose combinators for checking if a sequence has converged.

> If you are interested, have a look at this circularwindows example...

Oh that's really nice!

> Zip in Transducers.jl is more about fan-out than zipping so maybe it's just that I chose really bad name. Maybe I'll rename it to Tee or something.

Oh ok I'm still getting up to speed with the semantics but I'll watch out for this.

From your tl;dr I think I'll have a look at `eachcol` and `reshape`, maybe I can get things working with that approach.

Thanks for your help, and for all your work on this package!

**Comment by tkf** [comment]
Do you mean that it's useful to have `Zip` in the middle of a chain of transducers?  Or do you need to use some kind of moving window for calculating ELBO?

My comment was about moving window in particular and the way it appear in the code using transducers/iterators.  I do agree `Zip` is useful anywhere in a chain of transducers.

**Comment by cscherrer** [comment]
For `Zip`, I just need to figure out the semantics. Just some homework for me I think.

ELBO is my current use case, but I was thinking Transducers could be generally useful for separating concerns of specifying some sequence, vs reasoning about its convergence.

**Comment by tkf** [comment]
OK. Let me know if you find a case where `Transducers.Partition` is required and other things in the tl;dr didn't work. That would be a nice data point for me to think about how to optimize `Transducers.Partition`.

BTW, I'm also thinking about the interaction of RNG and Transducers #22. It's still at brainstorming phase, though.

I'm closing this issue for now and track the issue about `Zip` docs/name in #40.
